subsys/net/lib/dns/resolve.c

```diff
@@ -338,27 +339,102 @@ int dns_resolve_init(struct dns_resolve_context *ctx, const char *servers[],
 	return ret;
 }
 
+/* Check whether a slot is available for use, or optionally whether it can be
+ * reclaimed.
+ *
+ * @param pending_query the query slot in question
+ *
+ * @param reclaim_if_available if the slot is marked in use, but the query has
+ * been completed and the work item is no longer pending, complete the release
+ * of the slot.
+ *
+ * @return true if and only if the slot can be used for a new query.
+ */
+static inline bool check_query_active(struct dns_pending_query *pending_query,
+				      bool reclaim_if_available)
+{
+	int ret = false;
+
+	if (pending_query->cb != NULL) {
+		ret = true;
+		if (reclaim_if_available
+		    && pending_query->query == NULL
+		    && k_work_delayable_busy_get(&pending_query->timer) == 0) {
+			pending_query->cb = NULL;
+			ret = false;
+		}
+	}
+
+	return ret;
+}
+
+/* Must be invoked with context lock held */
 static inline int get_cb_slot(struct dns_resolve_context *ctx)
 {
 	int i;
 
 	for (i = 0; i < CONFIG_DNS_NUM_CONCUR_QUERIES; i++) {
-		if (!ctx->queries[i].cb) {
+		if (!check_query_active(&ctx->queries[i], true)) {
 			return i;
 		}
 	}
 
 	return -ENOENT;
 }
 
+/* Invoke the callback associated with a query slot, if still relevant.
+ *
+ * Must be invoked with context lock held.
+ *
+ * @param status the query status value
```

Comment:
nit: Param name missing?

Reply:
No, name changed to match the argument to  the function it's passed to, and I didn't update the implementation.  Fixed, also the missing one for the query pointer.